### ISSUE TITLE
Delay compaction when running out of disk space

### ISSUE THREAD

**Issue by nyh** [scylladb/scylla#2871]
Related to issue #2300 which is a more specific solution to the same problem.

Each compaction job has a known maximal temporary disk space needs, and especially in size-tiered compaction can be very large. If the amount of temporary disk needed exceeds the available disk space,  we should delay this compaction job until we have enough space. Moreover, we probably want to reserve at least some of the space for more pressing matters (like commit log, and perhaps flushing memtables) and not compactions.

Importantly, when we do start a compaction, we must "reserve" immediately the space it will need so other parallel compactions will not over-commit the same space. This is very important when we're running low on disk space, and need to avoid multiple shards doing huge compactions in parallel - in that case we need the different shards to compact one after another, instead of (if we're unlucky) starting all the compactions in parallel and using huge amounts of temporary disk space.

NOTE: If we do this feature, we can also run into a "deadlock" where we avoid a compaction because we're running low on disk space, but this compaction could have actually saved us space (i.e., when it removes deleted or overwritten data), and since we don't do it, we'll never use less space. Perhaps when we have low (but not yet critical) disk reserves, we should start giving more priority to compactions - especially those we guess will drop a lot of data - and less priority to write requests. This should probably have a separate issue.

**Comment by nyh** [comment]
The importance of this feature is significantly lower now that we have ICS (incremental compaction strategy) in all branches. Unlike STCS, ICS does not need to reserve potentially-huge amounts of disk space for the results of the compaction, because it compacts small pieces of an sstable at a time and deletes old pieces after creating new ones. 

There is still some temporary storage need, so compaction should not be attempted if the disk is 99% full, but at this point this has morphed into issue #3524, where if the disk is 99% full we need to decide which processes to do immediately to evict disk space, and which processes (like compaction) we should delay.

**Comment by mykaul** [comment]
https://github.com/scylladb/scylladb/issues/22806 seems to have the opposite view.
